proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -507,39 +507,1178 @@ locale delete_one = delete_one_conc + delete_one_abs +
+  apply (rule ccontr, clarsimp simp: obj_at_def)
+  apply (drule (3) reply_sc_refs)
+  apply (drule_tac sc=y in valid_replies_sc_with_reply_None)
+   apply (clarsimp simp: sc_at_ppred_def obj_at_def)+
+  done
+
+(* FIXME RT: update the one in Invariant_AI; we don't need valid_objs *)
+lemma reply_sc_refs':
+  "\<lbrakk>reply_sc reply = Some t; sym_refs (state_refs_of s);
+    kheap s rptr = Some (Structures_A.Reply reply)\<rbrakk>
+  \<Longrightarrow> \<exists>sc n. kheap s t = Some (Structures_A.SchedContext sc n) \<and>
+     (rptr \<in> set (sc_replies sc))"
+  apply (drule sym_refs_ko_atD[rotated])
+   apply (simp add: obj_at_def)
+  apply (clarsimp simp: get_refs_def2 obj_at_def valid_obj_def valid_reply_def refs_of_rev
+                        valid_bound_obj_def is_sc_obj_def state_refs_of_def image_iff)
+  done
+
+(* FIXME RT: replace the one in ArchFinalise_AI *)
+lemma sc_with_reply_None_reply_sc_reply_at:
+  "\<lbrakk> sc_with_reply rp s = None; reply_at rp s;
+      sym_refs (state_refs_of s); valid_replies s\<rbrakk>
+       \<Longrightarrow> reply_sc_reply_at ((=) None) rp s"
+  apply (rule ccontr, clarsimp simp: obj_at_def reply_sc_reply_at_def is_reply)
+  apply (drule not_sym, clarsimp)
+  apply (drule (2) reply_sc_refs')
+  apply (drule_tac sc=y in valid_replies_sc_with_reply_None, simp)
+   apply (clarsimp simp: sc_at_ppred_def obj_at_def)+
+  done
+
+(* replyClean *)
+
+crunches cleanReply
+ for valid_tcbs'[wp]: valid_tcbs'
+
+lemma no_fail_setReply [wp]:
+  "no_fail (reply_at' p) (setReply p reply)"
+  unfolding setReply_def
+  by (wpsimp simp: objBits_simps)
+
+lemma no_fail_updateReply [wp]:
+  "no_fail (reply_at' rp) (updateReply rp f)"
+  unfolding updateReply_def by wpsimp
+
+lemma no_fail_cleanReply [wp]:
+  "no_fail (reply_at' rp) (cleanReply rp)"
+  unfolding cleanReply_def
+  apply (rule no_fail_pre, rule no_fail_bind)
+     apply (wpsimp wp: updateReply_wp_all)+
+  apply (clarsimp simp: obj_at'_def ps_clear_upd' objBits_simps' projectKOs)
+  done
+
+lemma replySc_None_not_head:
+  "replySc reply = None \<longleftrightarrow> \<not> isHead (replyNext reply)"
+  by (fastforce simp: isHead_def getHeadScPtr_def split: reply_next.split_asm option.split_asm)
+
+(* reply_at_cross with property *)
+(* RT FIXME: use this to define reply_at_cross; do the same for other objects *)
+lemma reply_at_prop_cross:
+  assumes p: "pspace_relation (kheap s) (ksPSpace s')"
+  assumes ps: "pspace_aligned s" "pspace_distinct s"
+  assumes P: "\<And>r r'. reply_relation r r' \<Longrightarrow> P r \<Longrightarrow> P' r'"
+  assumes t: "obj_at (\<lambda>ko. \<exists>r. ko = Structures_A.Reply r \<and> P r) ptr s"
+  shows "obj_at' (\<lambda>r'::reply. P' r') ptr s'"
+  using assms
+  apply (clarsimp simp: obj_at_def is_reply)
+  apply (drule (1) pspace_relation_absD, clarsimp)
+  apply (case_tac z; simp)
+  by (fastforce dest!: aligned_distinct_obj_at'I[where 'a=reply] elim: obj_at'_weakenE)
+
+lemma reply_sc_reply_at_cross:
```

Comment:
Yes, we should consolidate on the projection version. I will clean this up. Some lemmas in this PR are written before you introduced `replySCs`, and I thought I've cleaned things up since, but apparently I've missed a few!